### ISSUE TITLE
20x slower training speed when updating to MMDetection 2 to 3 and SageMaker Deep Learning Container

### ISSUE THREAD

**Issue by matthost** [open-mmlab/mmdetection#11503]
I am seeing 20x slower training speed when updating MMDet and the SageMaker Deep Learning Container being used. Some colleagues are seeing similar slowness with different configs. Must be some issue with the docker build. Any idea to troubleshoot?

I am using [SageMaker Deep Learning Container](https://github.com/aws/deep-learning-containers/blob/master/available_images.md) with the following dockerfile:

```
FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:2.1.0-gpu-py310-cu121-ubuntu20.04-sagemaker

WORKDIR /opt/ml/code
ENV ENVROOT /opt/ml/code
COPY . /opt/ml/code

RUN pip install --upgrade pip

# install mmdetection and related dependencies per https://mmdetection.readthedocs.io/en/latest/get_started.html
RUN pip install -U openmim
RUN mim install mmengine
RUN mim install "mmcv>=2.0.0"
RUN mim install mmdet

ENV SAGEMAKER_SUBMIT_DIRECTORY /opt/ml/code
ENV SAGEMAKER_PROGRAM train.py

WORKDIR /
```

`train.py` is a torchrun launcher script like in https://github.com/open-mmlab/mmdetection/issues/8910

The mmdetection 2.x training dockerfile, which is 20x faster training time, is the following:

```
FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:1.9.1-gpu-py38-cu111-ubuntu20.04

WORKDIR /opt/ml/code
ENV ENVROOT /opt/ml/code
COPY . /opt/ml/code

RUN pip3 install --upgrade pip

# Install yapf before installing mmdet/mmcv-full. mmdet depends on mmcv which internally install yapf
# mmcv requirements do not tag any version for yapf and the latest version causes issues
# During the mmdet installation, yapf installation should be skipped
# https://github.com/open-mmlab/mmengine/issues/1480
RUN pip3 install yapf==0.32

RUN pip3 install tensorboard future tqdm==4.62.2
RUN FORCE_CUDA=1 pip3 install mmcv-full==1.4.5 -f https://download.openmmlab.com/mmcv/dist/cu111/torch1.9.1/index.html
RUN pip3 install mmdet==2.25.0

ENV SAGEMAKER_SUBMIT_DIRECTORY /opt/ml/code
ENV SAGEMAKER_PROGRAM train.py

WORKDIR /
```

I've tried to compile MMDetection 3.x with the torch 1.9 base image to try to isolate the issue, but it has an error related to missing `libtorch_cuda_cu.so` when importing `mmcv.ops.roi_align`. So far I have seen the same slowness issue with other more recent SageMaker DLC versions I've used, such as torch 2.0.

Since SageMaker Training is a common way to perform training it would be helpful to have guidance available for how to correctly assemble a container for it.

**Comment by matthost** [comment]
Thankfully SageMaker dropped a PyTorch 2.2 DLC and using that seems to have fixed the issue.

```
FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:2.2.0-gpu-py310-cu121-ubuntu20.04-sagemaker

WORKDIR /opt/ml/code
ENV ENVROOT /opt/ml/code
COPY . /opt/ml/code

ENV FORCE_CUDA=1
RUN pip install --upgrade pip

# install mmdetection and related dependencies per https://mmdetection.readthedocs.io/en/latest/get_started.html
RUN pip install -U openmim
RUN mim install mmengine
RUN mim install "mmcv>=2.0.0"
RUN mim install mmdet

ENV SAGEMAKER_SUBMIT_DIRECTORY /opt/ml/code
ENV SAGEMAKER_PROGRAM train.py

WORKDIR /

```


MMDet 2:
```

2024-02-27T21:37:49.611Z    2024-02-27 21:37:49,025 - mmdet - INFO - Epoch [1][50/13618]#011lr: 4.995e-06, eta: 5 days, 21:26:03, time: 1.039, data_time: 0.206, memory: 11731, loss_rpn_cls: 0.4977, loss_rpn_bbox: 0.0667, loss_cls: 2.8192, acc: 1.3858, loss_bbox: 0.2480, loss_mask: 2.4916, loss_mask_iou: 0.0250, loss: 6.1483

2024-02-27T21:38:20.620Z    2024-02-27 21:38:20,341 - mmdet - INFO - Epoch [1][100/13618]#011lr: 9.990e-06, eta: 4 days, 17:21:33, time: 0.626, data_time: 0.068, memory: 11731, loss_rpn_cls: 0.2133, loss_rpn_bbox: 0.0664, loss_cls: 1.8984, acc: 55.0860, loss_bbox: 0.3170, loss_mask: 1.1107, loss_mask_iou: 0.0136, loss: 3.6194

2024-02-27T21:38:53.629Z    2024-02-27 21:38:52,995 - mmdet - INFO - Epoch [1][150/13618]#011lr: 1.499e-05, eta: 4 days, 9:11:57, time: 0.653, data_time: 0.071, memory: 11731, loss_rpn_cls: 0.1053, loss_rpn_bbox: 0.0664, loss_cls: 0.6885, acc: 85.8110, loss_bbox: 0.4930, loss_mask: 0.6236, loss_mask_iou: 0.0116, loss: 1.9883

2024-02-27T21:39:25.637Z    2024-02-27 21:39:24,859 - mmdet - INFO - Epoch [1][200/13618]#011lr: 1.998e-05, eta: 4 days, 4:34:44, time: 0.637, data_time: 0.069, memory: 11731, loss_rpn_cls: 0.0659, loss_rpn_bbox: 0.0608, loss_cls: 0.5338, acc: 84.9388, loss_bbox: 0.5409, loss_mask: 0.4779, loss_mask_iou: 0.0101, loss: 1.6895

2024-02-27T21:39:57.646Z    2024-02-27 21:39:56,716 - mmdet - INFO - Epoch [1][250/13618]#011lr: 2.498e-05, eta: 4 days, 1:47:56, time: 0.637, data_time: 0.074, memory: 11731, loss_rpn_cls: 0.0531, loss_rpn_bbox: 0.0576, loss_cls: 0.4164, acc: 87.7798, loss_bbox: 0.4331, loss_mask: 0.3912, loss_mask_iou: 0.0100, loss: 1.3614

2024-02-27T21:40:28.655Z    2024-02-27 21:40:28,475 - mmdet - INFO - Epoch [1][300/13618]#011lr: 2.997e-05, eta: 3 days, 23:53:28, time: 0.635, data_time: 0.077, memory: 11731, loss_rpn_cls: 0.0436, loss_rpn_bbox: 0.0531, loss_cls: 0.3610, ac
```

MMDet 3 with torch 2.2 base:
```

2024-02-27T21:23:52.677Z    02/27 21:23:52 - mmengine - INFO - Epoch(train) [1][ 50/13618] lr: 1.9820e-05 eta: 4 days, 2:08:04 time: 0.7207 data_time: 0.1508 memory: 6906 loss: 3.8845 loss_rpn_cls: 0.4157 loss_rpn_bbox: 0.0807 loss_cls: 1.8160 acc: 88.2324 loss_bbox: 0.3748 loss_mask: 1.1564 loss_mask_iou: 0.0409

2024-02-27T21:24:34.689Z    02/27 21:24:34 - mmengine - INFO - Epoch(train) [1][ 100/13618] lr: 3.9840e-05 eta: 4 days, 9:38:53 time: 0.8312 data_time: 0.0890 memory: 7008 loss: 1.9530 loss_rpn_cls: 0.0597 loss_rpn_bbox: 0.0628 loss_cls: 0.6655 acc: 88.5742 loss_bbox: 0.6511 loss_mask: 0.5023 loss_mask_iou: 0.0117

2024-02-27T21:25:07.700Z    02/27 21:25:06 - mmengine - INFO - Epoch(train) [1][ 150/13618] lr: 5.9860e-05 eta: 4 days, 4:01:43 time: 0.6524 data_time: 0.0975 memory: 7008 loss: 1.2201 loss_rpn_cls: 0.0381 loss_rpn_bbox: 0.0542 loss_cls: 0.3806 acc: 93.2617 loss_bbox: 0.4225 loss_mask: 0.3148 loss_mask_iou: 0.0098

2024-02-27T21:25:37.710Z    02/27 21:25:37 - mmengine - INFO - Epoch(train) [1][ 200/13618] lr: 7.9880e-05 eta: 3 days, 23:57:55 time: 0.6157 data_time: 0.1011 memory: 7016 loss: 0.9170 loss_rpn_cls: 0.0289 loss_rpn_bbox: 0.0480 loss_cls: 0.3031 acc: 93.6035 loss_bbox: 0.3130 loss_mask: 0.2155 loss_mask_iou: 0.0085

2024-02-27T21:26:09.720Z    02/27 21:26:09 - mmengine - INFO - Epoch(train) [1][ 250/13618] lr: 9.9900e-05 eta: 3 days, 21:57:18 time: 0.6315 data_time: 0.1117 memory: 7008 loss: 0.7962 loss_rpn_cls: 0.0246 loss_rpn_bbox: 0.0473 loss_cls: 0.2615 acc: 95.2637 loss_bbox: 0.2813 loss_mask: 0.1737 loss_mask_iou: 0.0077

2024-02-27T21:26:41.730Z    02/27 21:26:41 - mmengine - INFO - Epoch(train) [1][ 300/13618] lr: 1.1992e-04 eta: 3 days, 21:00:19 time: 0.6488 data_time: 0.1188 memory: 7023 loss: 0.7826 loss_rpn_cls: 0.0272 loss_rpn_bbox: 0.0510 loss_cls: 0.254
```


MMDet 3 with torch 2.1 base:
`
02/27 21:47:44 - mmengine - INFO - Epoch(train) [1][ 50/13618] lr: 1.9820e-05 eta: 63 days, 12:51:04 time: 11.1985 data_time: 0.7748 memory: 6852 loss: 4.7638 loss_rpn_cls: 0.4010 loss_rpn_bbox: 0.0787 loss_cls: 2.1610 acc: 85.8887 loss_bbox: 0.2926 loss_mask: 1.7937 loss_mask_iou: 0.0368`
